### PR TITLE
Added support for TLS to MLMD GRPC Server

### PR DESCRIPTION
**Pull request by hbelmiro**
## The issue resolved by this Pull Request:
Resolves https://issues.redhat.com/browse/RHOAIENG-4971

This PR depends on:

- https://github.com/opendatahub-io/data-science-pipelines/pull/72

## Description of the changes:
Added support for TLS to MLMD GRPC Server

## Testing instructions

There are 2 scenarios for testing:

<details>
  <summary>TLS Enabled</summary>

1. Deploy the following DSPA

    ```yaml
    apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
    kind: DataSciencePipelinesApplication
    metadata:
      name: dspa
    spec:
      dspVersion: v2
      podToPodTLS: true
      apiServer:
        image: "quay.io/opendatahub/ds-pipelines-api-server:pr-72"
        argoDriverImage: "quay.io/opendatahub/ds-pipelines-driver:pr-72"
        argoLauncherImage: "quay.io/opendatahub/ds-pipelines-launcher:pr-72"
        enableSamplePipeline: true
      persistenceAgent:
        image: "quay.io/opendatahub/ds-pipelines-persistenceagent:pr-72"
      scheduledWorkflow:
        image: "quay.io/opendatahub/ds-pipelines-scheduledworkflow:pr-72"
      mlmd:  
        deploy: true  # Optional component
        grpc:
          image: "quay.io/opendatahub/mlmd-grpc-server:latest"
        envoy:
          image: "registry.redhat.io/openshift-service-mesh/proxyv2-rhel8:2.3.9-2"
      mlpipelineUI:
        deploy: true  # Optional component 
        image: "quay.io/opendatahub/ds-pipelines-frontend:pr-72"
      objectStorage:
        minio:
          deploy: true
          image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
    ```
2. Run the sample pipeline
3. The run must complete successfully
</details>

<details>
  <summary>TLS Disabled</summary>

1. Deploy the following DSPA

    ```yaml
    apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
    kind: DataSciencePipelinesApplication
    metadata:
      name: dspa
    spec:
      dspVersion: v2
      podToPodTLS: false
      apiServer:
        image: "quay.io/opendatahub/ds-pipelines-api-server:pr-72"
        argoDriverImage: "quay.io/opendatahub/ds-pipelines-driver:pr-72"
        argoLauncherImage: "quay.io/opendatahub/ds-pipelines-launcher:pr-72"
        enableSamplePipeline: true
      persistenceAgent:
        image: "quay.io/opendatahub/ds-pipelines-persistenceagent:pr-72"
      scheduledWorkflow:
        image: "quay.io/opendatahub/ds-pipelines-scheduledworkflow:pr-72"
      mlmd:  
        deploy: true  # Optional component
        grpc:
          image: "quay.io/opendatahub/mlmd-grpc-server:latest"
        envoy:
          image: "registry.redhat.io/openshift-service-mesh/proxyv2-rhel8:2.3.9-2"
      mlpipelineUI:
        deploy: true  # Optional component 
        image: "quay.io/opendatahub/ds-pipelines-frontend:pr-72"
      objectStorage:
        minio:
          deploy: true
          image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
    ```
2. Run the sample pipeline
3. The run must complete successfully
</details>

## Checklist
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by dsp-developers** [comment]
A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`
An OCP cluster where you are logged in as cluster admin is required.

To use this image run the following:

```bash
cd $(mktemp -d)
git clone git@github.com:opendatahub-io/data-science-pipelines-operator.git
cd data-science-pipelines-operator/
git fetch origin pull/683/head
git checkout -b pullrequest 85a6ca9ebb69f28e05780de54509b0d3855cb533
oc new-project opendatahub
make deploy IMG="quay.io/opendatahub/data-science-pipelines-operator:pr-683"
```

More instructions [here](https://github.com/opendatahub-io/data-science-pipelines-operator#deploy-dsp-instance) on how to deploy and test a Data Science Pipelines Application.

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

File: api/v1alpha1/dspipeline_types.go
Comment (gregsheremeta) [inline api/v1alpha1/dspipeline_types.go:348]: I don't love that the user needs to paste these in here. For apiserver, we are able to just use the paths to the OpenShift-provded certs like so:
https://github.com/opendatahub-io/data-science-pipelines-operator/blob/e101b0fc722048a6bcbdf4f2a7efa69281e71b73/config/internal/apiserver/default/deployment.yaml.tmpl#L180-L191

could we not do something similar for here?

Comment (hbelmiro) [reply api/v1alpha1/dspipeline_types.go:348]: This is automatically set by the operator and is not for users to use (technically they can). I had to add these fields to be able to set the contents in the template.
Regarding setting the contents, I also don't like it, but the MLMD config requires that.

Comment (gregsheremeta) [reply api/v1alpha1/dspipeline_types.go:348]: re: having to set the contents -- sure, no way around that.

> I had to add these fields to be able to set the contents in the template.

Ok, I understand the reason. I really don't like adding fields to the API for this though. Hmm. How sure are we that there's no other way to feed things to the manifestival templating? That doesn't sound right to me.

Comment (gregsheremeta) [reply api/v1alpha1/dspipeline_types.go:348]: yeah I think you could just add two strings roughly here
https://github.com/opendatahub-io/data-science-pipelines-operator/blob/20dd747d4342088a42dfce3cc98f907573ea3f82/controllers/dspipeline_params.go#L70

and then populate them

and then use those to feed the templating engine

and avoid changing the DSPA API

File: controllers/config/defaults.go
Comment (hbelmiro) [reply controllers/config/defaults.go:103]: We'll ignore the secret on tests. Since tests don't run on OpenShift, the secret is never created.
Should I add a comment?

File: controllers/dspipeline_params.go
Comment (hbelmiro) [reply controllers/dspipeline_params.go:907]: See https://github.com/opendatahub-io/data-science-pipelines-operator/pull/683#discussion_r1747267463.

File: controllers/mlmd.go
Comment (HumairAK) [inline controllers/mlmd.go:72]: Interesting, okay so I think you're doing this because we need the tls/key pair generated to submit into the protofile. One issue I see with this is that it could result in potential failures initially, because service-ca is creating these secrets asynchronously, and may not create it in time before we attempt to fetch it. 

I'm wondering if it might be better to do this in stages: 

* first reconcile and create the service
* ensure dspo triggers a reconcile when the service-ca tls secret is created 
* in the new reconcile, detect that we are podtopodtls == true && secretCreated == true, then proceed with r.ApplyDir(mlmddir)


Interested to hear thoughts, @gregsheremeta  / @hbelmiro

**Review by HumairAK — CHANGES_REQUESTED** [review]

File: controllers/mlmd.go
Comment (hbelmiro) [reply controllers/mlmd.go:72]: @HumairAK we would wait for the new reconcile indefinitely and never know if the secret was never created.
What about trying to `LoadMlmdCertificates` and repeat until a timeout (maybe 30s?)? It would be simpler and we can log an error if the certificates are not found.

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-683`

File: controllers/config/defaults.go
Comment (hbelmiro) [reply controllers/config/defaults.go:103]: We have tests that have TLS enabled, for example https://github.com/opendatahub-io/data-science-pipelines-operator/blob/cf1bd60823103ca99d4085e5f9c0a4c4e3fc3384/controllers/testdata/declarative/case_6/deploy/04_cr.yaml#L15

File: controllers/dspipeline_params.go
Comment (hbelmiro) [reply controllers/dspipeline_params.go]: Done.

File: config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl
Comment (hbelmiro) [reply config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl]: Done.

File: controllers/dspipeline_params.go
Comment (hbelmiro) [reply controllers/dspipeline_params.go]: Done.

File: api/v1alpha1/dspipeline_types.go
Comment (hbelmiro) [reply api/v1alpha1/dspipeline_types.go]: Done.

File: controllers/mlmd.go
Comment (hbelmiro) [reply controllers/mlmd.go:72]: > > @HumairAK we would wait for the new reconcile indefinitely and never know if the secret was never created.
> 
> why?

Because the second reconciliation would never be triggered.
I'm wondering now if there is any chance of the secrets never be created. This is automatically done by OpenShift, so we probably can trust it will always create them. WDYT?

> I think this would block the reconcile loop, which we don't want to do on a single dspa cr

That's a fair point.

Comment (HumairAK) [reply controllers/mlmd.go:72]: Because the second reconciliation would never be triggered.
I'm wondering now if there is any chance of the secrets never be created. This is automatically done by OpenShift, so we probably can trust it will always create them. WDYT? 

we can control how we reconcile based on the secret via https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/controllers/dspipeline_controller.go#L570, you can see that today we reconcile whenever a secret that the DSPA owns is updated/changed, we could trigger on any secret and filter on owned or service-ca secret 

> I'm wondering now if there is any chance of the secrets never be created. 
we could fall back on requeue and try again as well, but yeah if the secret is not created, then even the current implementation would probably not work

File: controllers/config/defaults.go
Comment (HumairAK) [reply controllers/config/defaults.go:103]: let's disable it then, and create an action item jira for adding integration tests for this, it will be blocked by https://issues.redhat.com/browse/RHOAIENG-11646 is done

Comment (hbelmiro) [reply controllers/config/defaults.go]: Test removed and Jira created: https://issues.redhat.com/browse/RHOAIENG-13123.

File: controllers/mlmd.go
Comment (hbelmiro) [reply controllers/mlmd.go]: Done.

**Comment by gmfrasca** [comment]
this is silly/trivial, but we should remove the UPSTREAM: <carry> prefixes in some of the commits messages/PR title... this repo *is* the upstream so we're not carrying anything here :)

File: controllers/util/util.go
Comment (gmfrasca) [inline controllers/util/util.go:202]: I feel we should make this generic, ie simply `GetSecret`, and take in the name of the secret as a parameter instead of hard-coding it here (especially since this in in `util.go`, which contains mostly generic helper functions)

File: controllers/mlmd.go
Comment (gmfrasca) [inline controllers/mlmd.go:80]: (this is really an improvement needed for `ApplyDir`, but until we fix it we should work around it)

The nomeclature of this method doesn't make it obvious that the functionality does _not_ recursively apply all subdirectories.  given that, we should move the files not in the `grpc-service` sudirectory into a seperate one to avoid confusion.  `default` seems to be the de-facto standard here (see other components such as apiserver, for example)

Comment (gmfrasca) [reply controllers/mlmd.go:80]: (related: created https://issues.redhat.com/browse/RHOAIENG-13196 to address above point)

File: controllers/dspipeline_params.go
Comment (gmfrasca) [inline controllers/dspipeline_params.go:598]: am i reading this right? ReplaceAll newline with the same newline seems like it shouldn't do anything

File: config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl
Comment (gmfrasca) [inline config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl:22]: small typo: `mTLS`

Comment (gregsheremeta) [reply config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl:22]: it actually IS for mTLS, which I don't think we're supporting in this PR ... but I'm not all the way through it yet :)

Ref: https://github.com/google/ml-metadata/blob/master/ml_metadata/proto/metadata_store.proto#L869-L870

"client certificate" i.e. mTLS vs regular TLS

Comment (gregsheremeta) [reply config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl:22]: well it's always hardcoded false, so no, we're not supporting mTLS here.

I'd have probably put
```suggestion
      client_verify: false // controls mTLS, which we don't use, so hardcode to false
```

File: controllers/util/util.go
Comment (gregsheremeta) [reply controllers/util/util.go:202]: +1. I believe that's what @HumairAK meant when he previously said

> this can be a static helper method [like we do for configmaps](https://github.com/opendatahub-io/data-science-pipelines-operator/blob/6f28be9771577deb0cc0cc7c12a4ca690635f900/controllers/util/util.go#L87-L99)

**Comment by gregsheremeta** [comment]
> we should remove the UPSTREAM: <carry> prefixes in some of the commits messages/PR title

+1, <carry> and friends is only applicable to DSP

File: config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl
Comment (gregsheremeta) [inline config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl:18]: I see that there's still database config in the deployment [here](https://github.com/opendatahub-io/data-science-pipelines-operator/pull/683/files#diff-a4e0c9838f36c9cbebe2ac4004d445a4bc30e6957441ef1783766dd588d74a73R27-R31).

Remind me why we have to have the db config here too? Are the deployment config settings ignored if this config.proto file is used? (If the answer is yes, I'd be tempted to move to always sending in config.proto and removing the deployment db config, but that's a tech debt problem for another day.)

File: controllers/mlmd.go
Comment (gregsheremeta) [inline controllers/mlmd.go:76]: I'm not an expert on this, but I would expect to see a more of a warning and a re-reconcile, not an error. I think the error might look a little more scary in the logs as well. I'd like Humair's opinion though.

File: controllers/mlmd.go
Comment (gregsheremeta) [inline controllers/mlmd.go:63]: Why did we need to move the two service templates to their own subdirectory? Is it so that we can trick OpenShift into giving us our certs first? (The services are what have that annotation to make it hand over the certs.)

Might be worth a comment.

File: controllers/mlmd.go
Comment (gregsheremeta) [reply controllers/mlmd.go:80]: I think he's actually taking advantage of the fact that ApplyDir doesn't recurse. I'm pretty sure he wants the services to go out first on their own. You might then have feelings about using a subdir here if you want to fix up ApplyDir :)

**Review by gregsheremeta — COMMENTED** [review]
lgtm after a few nitpicks and questions are addressed. Very nice!

File: controllers/mlmd.go
Comment (gregsheremeta) [reply controllers/mlmd.go]: reviewed this logic in the PR, and it looks great.

File: api/v1alpha1/dspipeline_types.go
Comment (gregsheremeta) [reply api/v1alpha1/dspipeline_types.go]: looks perfect!

File: controllers/dspipeline_params.go
Comment (hbelmiro) [reply controllers/dspipeline_params.go:598]: It's not the same newline. I'm replacing it with an escaped `\`.

File: config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl
Comment (hbelmiro) [reply config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl:18]: Actually, this config is not even used. I had to add it because `connection_config` is mandatory. It doesn't make any difference if we set the values to `unused`, for example.

File: controllers/mlmd.go
Comment (hbelmiro) [reply controllers/mlmd.go:76]: I agree. But an error is necessary to update the DSPA status with `not ready`.
Maybe we should improve the DSPA status mechanism to handle these situations.

@HumairAK thoughts?

File: controllers/util/util.go
Comment (hbelmiro) [reply controllers/util/util.go]: Done.

File: config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl
Comment (hbelmiro) [reply config/internal/ml-metadata/metadata-grpc-tls-config-secret.yaml.tmpl]: Comment changed as suggested.

File: controllers/mlmd.go
Comment (hbelmiro) [reply controllers/mlmd.go:64]: Done.

**Comment by hbelmiro** [comment]
@gmfrasca 

> this is silly/trivial, but we should remove the `UPSTREAM: <carry>` prefixes in some of the commits messages/PR title... this repo _is_ the upstream so we're not carrying anything here :)

Removed and commits squashed.

File: controllers/mlmd.go
Comment (HumairAK) [reply controllers/mlmd.go:77]: yeah we don't have a protocol for handling readying state, this will have to do for now

I think the solution is relatively simple, just have the independent component reconcilers functions return a status condition/message, and put that for ready status, but this can be it's own separate Jira, @hbelmiro can you create one?

File: controllers/testdata/declarative/case_6/deploy/04_cr.yaml
Comment (HumairAK) [inline controllers/testdata/declarative/case_6/deploy/04_cr.yaml:8]: this test was doing more than just testing podtopodtls, but I don't want to block this pr on re-adding this back in, but we should not forget to re-add this once the pr is merged, as a separate task item

**Review by HumairAK — CHANGES_REQUESTED** [review]

File: controllers/mlmd.go
Comment (hbelmiro) [reply controllers/mlmd.go:77]: https://issues.redhat.com/browse/RHOAIENG-13321

File: controllers/testdata/declarative/case_6/deploy/04_cr.yaml
Comment (hbelmiro) [reply controllers/testdata/declarative/case_6/deploy/04_cr.yaml:8]: I created a Jira for that: https://issues.redhat.com/browse/RHOAIENG-13123

File: controllers/dspipeline_controller.go
Comment (hbelmiro) [reply controllers/dspipeline_controller.go]: Makes sense.
Done.

File: controllers/dspipeline_controller.go
Comment (hbelmiro) [reply controllers/dspipeline_controller.go:313]: Done.

**Pull request merged**
